homeassistant/components/tedee/coordinator.py

```diff
@@ -0,0 +1,110 @@
+"""Coordinator for Tedee locks."""
+import asyncio
+from datetime import timedelta
+import logging
+import time
+
+from pytedee_async import (
+    TedeeAuthException,
+    TedeeClient,
+    TedeeClientException,
+    TedeeDataUpdateException,
+    TedeeLocalAuthException,
+    TedeeLock,
+)
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import CONF_ACCESS_TOKEN, CONF_HOST
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import ConfigEntryAuthFailed
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed
+
+from .const import CONF_LOCAL_ACCESS_TOKEN, DOMAIN
+
+SCAN_INTERVAL = timedelta(seconds=20)
+STALE_DATA_INTERVAL = 300
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class TedeeApiCoordinator(DataUpdateCoordinator[dict[int, TedeeLock]]):
+    """Class to handle fetching data from the tedee API centrally."""
+
+    def __init__(self, hass: HomeAssistant, entry: ConfigEntry) -> None:
+        """Initialize coordinator."""
+        self.tedee_client = TedeeClient(
+            personal_token=entry.data.get(CONF_ACCESS_TOKEN),
+            local_token=entry.data.get(CONF_LOCAL_ACCESS_TOKEN),
+            local_ip=entry.data.get(CONF_HOST),
+        )
+        self._initialized = False
+        self._next_get_locks = time.time()
+        self._last_data_update = time.time()
+        self._stale_data = False
+        super().__init__(
+            hass,
+            _LOGGER,
+            name=DOMAIN,
+            update_interval=SCAN_INTERVAL,
+        )
+
+    async def _async_update_data(self) -> dict[int, TedeeLock]:
+        """Fetch data from API endpoint."""
+
+        try:
+            _LOGGER.debug("Update coordinator: Getting locks from API")
+
+            # once every hours get all lock details, otherwise use the sync endpoint
+            if self._next_get_locks - time.time() <= 0:
+                _LOGGER.debug("Updating through /my/lock endpoint")
+                await self.tedee_client.get_locks()
+                self._next_get_locks = time.time() + 60 * 60
+            else:
+                _LOGGER.debug("Updating through /sync endpoint")
+                await self.tedee_client.sync()
+
+            self._last_data_update = time.time()
+
+        except TedeeLocalAuthException as ex:
+            raise ConfigEntryAuthFailed(
```

Comment:
Please raise `ConfigEntryError` for now as you don't have reauth flow in this PR